personalrobotics/ada_feeding
# Add acquisition context, posthoc, and policy service

# Description

(1) Adds the AcquisitionSchema Message, which holds an instance of a CoRL2023 action

(2) Adds the services AcquisitionSelect and AcquisitionReport.

The former is designed to take in the Mask (incl. ROI) associated with the desired food item and return the action we want to execute, the latter is used to report the result of that action (plus any posthoc context) to the acquisition policy.

(3) Add the package `ada_feeding_actionselect` to implement the above services. This incorporates: [a] a Context Adapter that converts the incoming Mask + image (retrieved via topic subscription) into a single context vector, [b] the actual policy that returns a _discrete_ action distribution (i.e. a list of actions + a parallel list of probabilities), and [c] a Posthoc Adapter that can optionally change the incoming posthoc into a new vector.

Currently, the only Policy implemented is `ConstantPolicy` that always returns a single action w/ probability 1.
Currently, the only Context/Posthoc adapter implemented is `NoContext`, which completely drops the image/haptics and just send `[0.0]` to the policy.

This PR leaves a lot of TODOs, including (in order of importance):
* Adding SPANet as a Context Adapter
* Adding HapticNet and Passthrough as Posthoc Adapters
* Adding Contextual Bandits as Policies
* Adding a policy cache for learned policies (so we don't start from scratch every time)
* Adding Data collection (i.e. record the mask, image, context vector, posthoc vector, action distribution, action taken, and loss)

# Testing procedure

* In `policies.yaml`, edit `constant.kwargs.index` with the exact action in `acquisition_library.yaml` you want to get
* Run the service (defaulting to the `constant` policy): `ros2 launch ada_feeding_actionselect ada_feeding_actionselect_launch.xml`
* In a separate terminal, call action_select: `ros2 service call /ada_feeding_actionselect/action_select ada_feeding_msgs/srv/AcquisitionSelect`. Verify that the response matches the schema action from the xml.
* In that terminal, call action_report: `ros2 service call /ada_feeding_actionselect/action_report ada_feeding_msgs/srv/AcquisitionReport`, verify the response is a success.

Unfortunately, this PR also starts with the skeleton of the AcquireFood tree, so it is imperative to make sure this doesn't break anything (the YAML should still point to the dummy MoveTo tree for AcquireFood).

* Make sure this doesn't break anything.

# Before opening a pull request

# Before Merging
